# Contents List

## 01. Importing Libraries
## 02. Paths
## 03. Importing Data
## 04. Data Consistency Checks
## 05. Task-4.5

# 01. Importing libraries

In [235]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Paths

In [238]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [240]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [243]:
#Import the products Dataset
df_prods = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'products.csv'), index_col = False)

In [245]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [247]:
#Import the orders Dataset
df_ords = pd.read_csv(os.path.join(path, '02-Data', 'Prepared-Data', 'orders_wrangled.csv'), index_col = False)

In [248]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# 04. Data Consistency Checks
### 01. Mixed-Type Data

In [250]:
#Creating a mix data type column in a test dataframe
df_test = pd.DataFrame()

In [251]:
#The mixed column if fille with different types of data 
df_test['mix'] = ['a', 'b', 1, True]

In [256]:
#Creating fuction to search for a mixed data type in a column
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_29147/2911751512.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [258]:
#Converting the mixed column data type to String as most of the data in the mixed column is of a string.
df_test['mix'] = df_test['mix'].astype('str')

In [260]:
df_test['mix'].dtype

dtype('O')

### 02. Missing Values

In [263]:
#Finding missing values in the products dataset
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [265]:
#Creating a datafarme to subset the data in the df_prods where there  missing values are found  in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True ]

In [267]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [269]:
#Running explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [271]:
#Fidning the median of the df_prods dataframe
df_prods['prices'].median() 

7.1

In [273]:
#Finding outliers in the df_prods dataframe.
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [275]:
#Replace the outliers values in the prices column with the meadian value of 7.1.
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [277]:
#Checking the df_prods dataframe after imputation
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

### Observation: Explanatory Analysis (Products Dataframe)
##### After running explanatory analysis on the df_products dataframe, it was discovered that there are outliers present in the data.
##### The prices column has a mean of 9.99 and a standard deviation of 453.6. This SD is more than 2 standard deviations away from the mean, indicating the presence of outlier values in the data.
##### FIdentifying all data in the prices column that is 2 standard deviations away from the mean resulted in two entries or rows: prices for 14900 and prices for 99999. 
##### These were identified as outliers and were subsequently replaced with the median value of 7.1. 
##### Given that the minimum value of 1 and maximum value of 99999 produce a wide range of values, it is recommended to replace outliers with the median value.
##### The median is less sensitive to outliers compared to the mean, making it a more appropriate measure for these extreme values. 
##### After replacing the outliers with the median value, the mean of the dataframe decreased to 7.6.

In [280]:
#Checking the number of rows & columns in df_prods dataframe
df_prods.shape

(49693, 5)

In [282]:
#Creating a new dataframe as a subset of df_prods dataframe where the missing values in the product_name variable are excluded
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [284]:
df_prods_clean.shape

(49677, 5)

### 03. Duplicate Values

In [287]:
#Fidning duplicates in the df_prods_clean dataframe and subset those duplicate data and store them in the df_dups -new dataframe 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [289]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [291]:
#Checking the number of rows in the df_prods_clean datafrane
df_prods_clean.shape

(49677, 5)

In [293]:
#Creating a new datafrme where there is no duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [295]:
df_prods_clean_no_dups.shape

(49672, 5)

# 05. Task-4.5

### This section focuses on the Orders dataframe
##### The dataframe has already been imported above
# Data Consistency Checks

In [299]:
# Running explanatory analysis in the df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### Observation 2. Explanatory Analysis
##### Investigating the accuracy of the columns in the df_ords dataframe. After examining the values in each column, no unreasonable values were found. All values in the columns appear to fall within an acceptable and interpretable range for their corresponding columns/variables.

In [302]:
#Checking the dataframe for any mixed type data within a column/variable in the df_ords dataframe
#Check for mixed-type data in your df_ords dataframe
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

### Observation 3. Mixed-Typed Data 
##### There is no mixed-type data found in any of the columns/variables in the df_ords dataframe.

In [305]:
# Running the fucntion to search for a mixed data type in a column in the df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)      

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_29147/1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_29147/1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_29147/1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_29147/1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

In [307]:
#Checking the data types of all the columns in the df_ords dataframe
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


### Observation 4. Mixed-Typed Data
##### There is no mixed-type data found in any of the columns/variables in the df_ords dataframe.
#####  This was  confirmed by searching for mixed-type data in a column using the a function and the result was NULL.
##### The info function also confirmed that no mixed-type data is present.
##### No mixed type data column/ variable was found. 

In [310]:
#Run a check for missing values in the df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [312]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [314]:
#Find out the the missing values based on user_id count summarized 
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

Unnamed: 0  order_id  order_number  orders_day_of_week  \
count    206209.0  206209.0      206209.0            206209.0   
mean          1.0       1.0           1.0                 1.0   
std           0.0       0.0           0.0                 0.0   
min           1.0       1.0           1.0                 1.0   
25%           1.0       1.0           1.0                 1.0   
50%           1.0       1.0           1.0                 1.0   
75%           1.0       1.0           1.0                 1.0   
max           1.0       1.0           1.0                 1.0   

       order_hour_of_day  days_since_prior_order  
count           206209.0                206209.0  
mean                 1.0                     0.0  
std                  0.0                     0.0  
min                  1.0                     0.0  
25%                  1.0                     0.0  
50%                  1.0                     0.0  
75%                  1.0                     0.0  
max                  1.0                     0.0

### Observation 5. Missing Values
##### There are a total of 206209 missing values that were found in the days_since_prior_order column/variable in the original df_ords dataframe. 
##### While the df_ords dataframe has a total of 3421083 rows and 7 columns, the total number of unique user_id is 206209. 
##### This suggests that each unique user_id has one missing value in the days_since_prior_order column, and that missing value belongs to the very first order of each unique user_id. 
##### This makes sense because there is no value/day before each first order.
##### And for this reason, it is unnecessary to remove these missing values.

### Observation 6. Missing Values

##### There is no need to remove these missing values, since each unique user_id should have only one missing value in the days_since_prior_order column, and that missing value belongs to the very first order of each unique user_id.


In [318]:
#Finding duplicates in the df_ords dataframe
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### Observation 7 & 8. Duplicates 
##### There are no duplicate values found in the df_ords dataframe.

In [321]:
#Chekcing the df_ords dataframe
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [323]:
#Dropping the column "Unnamed: 0" from the df_ords dataframe and assinged it to a new dataframe "df_ords_clean"
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [325]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### Observation - Extra
##### After checking the df_ords dataframe, it was noticed that there is an unnecessary column called "Unnamed: 0" that needs to be dropped. 

In [328]:
#Export the cleaned datafrme for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'products_checked.csv'))

In [330]:
#Export the cleaned datafrme for df_ords
df_ords_clean.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'orders_checked.csv'))

### Observation 9. Export

##### df_prods: final clean version = df_prods_clean_no_dups, and saved as products_checked.csv
##### df_ords: final clean version = df_ords_clean, and saved as orders_checked.csv